src/backend/access/transam/xact.c

```diff
@@ -7561,3 +7567,49 @@ MarkSubTransactionAssigned(void)
 
 	CurrentTransactionState->assigned = true;
 }
+
+
+void
+txn_stack_get_xid_walker(txn_stack_get_xid_callback func, void *arg)
+{
+	/* array of transaction stack xid */
+	for (TransactionState cur = CurrentTransactionState; cur != NULL; cur = cur->parent)
+	{
+		func(arg, cur->nestingLevel, cur->fullTransactionId);
+	}
+}
+
+void
+txn_stack_set_xid_walker(txn_stack_set_xid_callback func, void *arg)
+{
+	/* array of transaction stack xid */
+	for (TransactionState cur = CurrentTransactionState; cur != NULL; cur = cur->parent)
+	{
+		FullTransactionId savedFullXid = cur->fullTransactionId;
+
+		func(arg, cur->nestingLevel, &cur->fullTransactionId);
```

Comment:
design return value to judge for `continue` or not ?

Reply:
Move the test into the callback seems worth nothing :-)

Reply:
means we can control the callback flow.

Reply:
The function of the walker is very clear. So I prefer the walker undertakes more, otherwise every callback needs to repeate identical logic.